File: coding/src/com/algos/core/li10_tree/bt/revised/Bt_L104_MaximumDepthOfBinaryTree.java
package  com.algos.core.li10_tree.bt.revised;

import  com.algos.core.li30_model.TreeNode;

public class Bt_L104_MaximumDepthOfBinaryTree {

	public static void main(String[] args) {
		Bt_L104_MaximumDepthOfBinaryTree obj = new Bt_L104_MaximumDepthOfBinaryTree();
				
		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
		
		System.out.println(obj.maxDepth(root));
		
	}

	// Problem: https://leetcode.com/problems/maximum-depth-of-binary-tree/
	// Idea: Maximum depth is the number of nodes along the longest path to any leaf node.
	// 1 + max(left, right) 
	// Time: O(n)
    // space: O(1)
	public int maxDepth(TreeNode root) {
        if(root == null) 
			return 0;
		
		int leftDepth = maxDepth(root.left);
		int rightDepth = maxDepth(root.right);
		return 1 + Math.max(leftDepth, rightDepth);
    }
}

File: dsa/concepts/10_trees/bt/python/bt_l104__maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/maximum-depth-of-binary-tree/
# Idea: Maximum depth is the number of nodes along the longest path to any leaf node.
# 1 + max(left_depth, right_depth)
# Time: O(n)
# Space: O(1) ignoring recursion stack

def max_depth(root):
    if not root:
        return 0

    left_depth = max_depth(root.left)
    right_depth = max_depth(root.right)
    return 1 + max(left_depth, right_depth)


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(max_depth(root))  # Expected: 4
